Add Unit Test Support for StageMode Interaction

Phipson opened the pull request:
#### 0a8af7554b376e95b170d006e17816e4a7b4878d
<pre>
Add Unit Test Support for StageMode Interaction
<a href="https://bugs.webkit.org/show_bug.cgi?id=287943">https://bugs.webkit.org/show_bug.cgi?id=287943</a>
<a href="https://rdar.apple.com/145113469">rdar://145113469</a>

Reviewed by Wenson Hsieh and Ada Chan.

This PR adds support for unit tests for stagemode interaction. It leverages existing APIs for Drag and Drop
but adds additional calls to simulate the pan gesture recognizer for model interaction. Subsequent PRs will
be made to add additional support for testing correct behavior when mapping XY translation to rotation.

* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _simulatePanGestureBeginAtPoint:]):
(-[WKWebView _simulatePanGestureUpdateAtPoint:]):
(-[WKWebView _stageModeInfoForTesting]):
- Adds calls into the WKContentView to start and update the pan gesture
- Also adds a call to fetch information about the stageModeSession packaged as a dictionary

* Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
(TestWebKitAPI::TEST(DragAndDropTests, IgnoreHitTestStageModeModel)):
(TestWebKitAPI::TEST(DragAndDropTests, CanHitTestStageModeModel)):
- Adds unit tests to verify whether the returned information for stage mode is correct based on hit-testing

* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:
* Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm:
(-[DragAndDropSimulator hitTestForStageModeAt:]):
(-[DragAndDropSimulator awaitingStageModeHitResult]):
(-[DragAndDropSimulator stageModeHitTestValidModel]):
- Adds additional APIs to the simulator to call into the WKWebView&apos;s testing API to simulate a pan

Canonical link: <a href="https://commits.webkit.org/291054@main">https://commits.webkit.org/291054@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24159a52e546086032f3f6333b497c17cd5da007

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91325 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10859 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/353 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96297 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42032 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93375 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11237 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19176 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70121 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27650 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94326 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8559 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82712 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50447 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8328 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/293 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41182 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78649 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/303 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98293 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18492 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13552 "Too many flaky failures: fast/mediastream/MediaStream-removeTrack-while-playing.html, fast/mediastream/camera-unknown-facing-mode.html, fast/mediastream/cloned-video-stream-aspect-ratio.html, fast/mediastream/getUserMedia-to-canvas-1.html, fast/mediastream/getUserMedia-to-canvas-2.html, http/tests/pdf/page-in-window-update-with-linearized-pdf-in-display-none-iframe.html, imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page.html, imported/w3c/web-platform-tests/file-system-access/getDirectory.https.any.worker.html, media/now-playing-status-for-video-conference-web-page.html, webrtc/vp8-then-h264-gpu-process-crash.html, webrtc/vp9-sw.html, workers/sab/growable-shared-array-buffer-serialization.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79143 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18748 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78547 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78345 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19490 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22869 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/222 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11653 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18491 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23783 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18207 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21665 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19975 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -14352,6 +14352,26 @@ - (void)_dismissContactPickerWithContacts:(NSArray *)contacts
+- (NSDictionary *)_stageModeInfoForTesting
+{
+    return @{
+        @"awaitingResult" : @(_stageModeSession->isPreparingForInteraction),
```
`_stageModeSession` is an optional. Should we return empty dictionary early if it's null?

achan00 reviewed (approved):
Looks good if you can address the `_stageModeSession` comment.

Phipson committed:
Add Unit Test Support for StageMode Interaction
https://bugs.webkit.org/show_bug.cgi?id=287943
rdar://145113469

Reviewed by Wenson Hsieh and Ada Chan.

This PR adds support for unit tests for stagemode interaction. It leverages existing APIs for Drag and Drop
but adds additional calls to simulate the pan gesture recognizer for model interaction. Subsequent PRs will
be made to add additional support for testing correct behavior when mapping XY translation to rotation.

* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _simulatePanGestureBeginAtPoint:]):
(-[WKWebView _simulatePanGestureUpdateAtPoint:]):
(-[WKWebView _stageModeInfoForTesting]):
- Adds calls into the WKContentView to start and update the pan gesture
- Also adds a call to fetch information about the stageModeSession packaged as a dictionary

* Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
(TestWebKitAPI::TEST(DragAndDropTests, IgnoreHitTestStageModeModel)):
(TestWebKitAPI::TEST(DragAndDropTests, CanHitTestStageModeModel)):
- Adds unit tests to verify whether the returned information for stage mode is correct based on hit-testing

* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:
* Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm:
(-[DragAndDropSimulator hitTestForStageModeAt:]):
(-[DragAndDropSimulator awaitingStageModeHitResult]):
(-[DragAndDropSimulator stageModeHitTestValidModel]):
- Adds additional APIs to the simulator to call into the WKWebView's testing API to simulate a pan

Canonical link: https://commits.webkit.org/291054@main
